_overviews/tutorials/binary-compatibility-for-library-authors.md

```diff
@@ -0,0 +1,140 @@
+
+![Binary incompatibility after upgrading]({{ site.baseurl }}/resources/images/library-author-guide/after_update.png){: style="width: 280px; margin: auto; display: block;"}
+
+Why did we get a `NoSuchMethodError`? Remember that `A v1.0.0` is compiled with `C v1.0.0` and thus calls methods available in `Cv1.0.0`. While `B` and
+our App has been recompiled with available classes/methods in `C v2.0.0`, `A v1.0.0`'s bytecode hasn't changed - it still calls the same method that is now missing in `C v2.0.0`!
+
+This situation can only be resolved by ensuring that the chosen version of `C` is binary compatible with all other evicted versions of `C`. In this case, we need a new version of `A` that depends
+on `C v2.0.0` (or any other future `C` version that is binary compatible with `C v2.0.0`).
+
+Now imagine if our App is more complex with lots of dependencies themselves depending on `C` (either directly or transitively) - it becomes extremely difficult to upgrade any dependencies because it now
+pulls in a version of `C` that is incompatible with the rest of the versions of `C` in our dependency tree! In the example below, we cannot upgrade `D` because it will transitively pull in `C v2.0.0`, causing breakages
+due to binary incompatibility. This inability to upgrade any packages without breaking anything is common known as **Dependency Hell**.
+
+![Dependency Hell]({{ site.baseurl }}/resources/images/library-author-guide/dependency_hell.png)
+
+How can we, as library authors, spare our users of runtime errors and dependency hell?
+
+* Use **Migration Manager** (MiMa) to catch unintended binary compatibility breakages before releasing a new library version
+* **Avoid breaking binary compatibility** through careful design and evolution of your library interfaces
+* Communicate binary compatibility breakages clearly through **versioning**
+
+## MiMa - Check Binary Compatibility with Previous Library Versions 
+
+The [Migration Manager for Scala](https://github.com/typesafehub/migration-manager) (MiMa) is a tool for diagnosing binary incompatibilities between different library versions.
+
+When run standalone in the command line, it will compare the .class files in the two provided JARs and report any binary incompatibilities found. Most library authors use the [SBT plugin](https://github.com/typesafehub/migration-manager/wiki/Sbt-plugin)
+to help spot binary incompatibility between library releases. (Follow the link for instructions on how to use it in your project)
+
+## Designing for Evolution - without breaking binary compatibility
+
+TODO
+
+## Versioning Scheme - Communicate binary compatibility breakages
+
+We recommend using the following schemes to communicate binary and source compatibility to your users:
+
+* Any release with the same major version are **Binary Backwards Compatible** with each other
```

Comment:
Okay, this is how I propose you write this:

* Any library versioned with a starting `0` can break binary and source compatibility at any time, as of the [semantic versioning](link here) spec.
* A change in the **major version number** signals at least one **binary incompatibility**, either forwards and/or backwards. For example, code compiled against `v1.0.0` is not compatible with `v2.0.0`.
* A change in the **minor version number** signals at least one **source incompatibility**, either forwards and/or backwards. For example, code compiled against the sources of a dependency `v1.0.0` is not ensure to compile with `v1.1.0` of the same library.
* A change in the patch version (the `x` in `1.0.x`) is both binary and source compatible. Patch releases often include bug fixes and minor behavioural changes.
  
There may be cases in which libraries break source compatibility and decide to bump up the major version number. These cases, where not common, are defined per library.

Reply:
I think the new format is much more readable and more "structured". The dumber the bullet lists are, the better ^^

Reply:
One thing is not clear to me is the status we give to forward binary compatibility. We should probably say that it's often the case that libraries do not promise forward binary compatibility, and that therefore they don't influence major version numbers. I'll think about this in the next hours since now I'm saturated, but perhaps @sjrd can shed some light on this. He's the original author of this proposal.

Reply:
Forward compat is generally not an issue and does not count per se in the versioning number I am advocating.

However, checking forward binary compatibility with MiMa is the best approximation we have of guaranteed (100%) *backward source* compat. Indeed, the most common source for backward source incompatibility are additions of public fields/methods (which influence implicit resolution), and that is forbidden if you enforce forward binary compat. Therefore, I would in general encourage checking forward bin compat across patch releases (so between 1.2.x and 1.2.y), which is the best thing we have to enforce backward source compat in patch releases.

Reply:
* I agree with @sjrd regarding forward compatibility not being as relevant for most userland libraries
* I am still on the fence about minor versions. I know a lot of libraries like to use minor version to signal new features. Perhaps we make minor version a mandatory bump if forward compatibility is broken? (which normally coincides with adding new functionality like new methods).
* Does the SBT MiMa plugin currently support easy checking for forward compatibility?
* Just an idea: I think we should extend MiMa SBT plugin to automatically suggest a new version number based on the result of compatibility check. (Although I'm not sure if there's a way to check source compatibility..)